doc/adr/0003-distributed-tracing.md

```diff
@@ -0,0 +1,158 @@
+# 3. NATS Service Latency Distributed Tracing Interoperability
+
+Date: 2020-05-21
+Author: @ripienaar
+
+## Status
+
+Proposed
+
+## Context
+
+The goal is to enable the NATS internal latencies to be exported to distributed tracing systems, here we see a small
+architecture using Traefik, a Go microservice and a NATS hosted service all being observed in Jaeger.
+
+![Jaeger](0003-jaeger-trace.png)
+
+The lowest 3 spans were created from a NATS latency Advisory.
+
+These traces can be ingested by many other commercial systems like Data Dog and Honeycomb where they can augment the
+existing operations tooling in use by our users. Additionally Grafana 7 supports Jaeger and Zipkin today.  
+
+Long term I think every server that handles a message should emit a unique trace so we can also get visibility into
+the internal flow of the NATS system and exactly which gateway connection has a delay - see our current HM issues - but
```

Comment:
OK, our client latencies are also a bit of a lie though since it's all based on RTT and we do not update RTT enough to be useful.

But, if each server did send a span - spans include start and end times - we would see a gap between the time that gw1 sent the message to the TCP stack and the time gw2 says it received the message, this could be expressed as a RTT between 2 gateway endpoints so you'd get a good view from this, I think?

Reply:
If their clocks are synced. This smth we should start to check and warn about?

Reply:
It's where they are today, I am only proposing a small amendment to the existing advisories and model - larger changes like the non cross account ones is out of scope for this.

Reply:
* Since the measurement is eventually triggered by a header, I think it should be a system or operator setting rather than an export setting. It's not like not having set the measurement prevents it from being taken...
* Compared to the other changes proposed here, I don't see exporting different metrics via the system account as too big of a deal. In the long run, I'd be more concerned about having to support two kinds. (header is set, which kind to pick)

Reply:
I think it's reasonable that someone might want to say a particular service - perhaps due to content of the headers or because its somehow public/hostile facing - should opt out of these advisories.  Traditionally Derek has been quite against more and more advisories too :)

I like setting which system we're part of so we dont suddenly start trusting the wrong kind of header by accident of evaluation order, if its a jaeger traced export then thats all we consider.

Reply:
Since this thread has grown quite a bit sounds like time for a quick video call between interested parties.

Reply:
I sent an invite.

Reply:
However this is all a distraction and off topic :) We could resolve this one by just focussing on the question at hand.  Adding a feature to existing advisory behaviour does not require a refactor-the-world discussion